base_layer/wallet/src/output_manager_service/storage/database.rs

```diff
@@ -0,0 +1,376 @@
+                o.spending_key.clone(),
+                Box::new(o.clone()),
+            )))?;
+        }
+
+        let _ = self
+            .db
+            .write(WriteOperation::Remove(DbKey::PendingTransactionOutputs(tx_id)))?;
+
+        Ok(())
+    }
+
+    /// This method is called when a transaction is built to be sent. It will encumber unspent outputs against a pending
+    /// transaction
+    pub fn move_unspent_outputs_to_pending_transaction(
+        &mut self,
+        tx_id: TxId,
+        outputs_to_move: Vec<UnblindedOutput>,
+        change_output: Option<UnblindedOutput>,
+    ) -> Result<(), OutputManagerStorageError>
+    {
+        let mut outputs_to_be_spent = Vec::new();
+        for i in outputs_to_move {
+            if !self.db.contains(&DbKey::UnspentOutput(i.spending_key.clone()))? {
+                return Err(OutputManagerStorageError::ValuesNotFound);
+            }
+            match self
+                .db
+                .write(WriteOperation::Remove(DbKey::UnspentOutput(i.spending_key.clone())))?
+                .ok_or(OutputManagerStorageError::ValueNotFound(DbKey::UnspentOutput(
+                    i.spending_key.clone(),
+                )))? {
+                DbValue::UnspentOutput(o) => outputs_to_be_spent.push(*o),
+                _ => {
+                    return Err(OutputManagerStorageError::UnexpectedResult(
+                        "Incorrect returned type".to_string(),
+                    ))
+                },
+            }
+        }
+
+        let mut pending_transaction = PendingTransactionOutputs {
+            tx_id: tx_id.clone(),
+            outputs_to_be_spent,
+            outputs_to_be_received: Vec::new(),
+            timestamp: Utc::now().naive_utc(),
+        };
+
+        if let Some(co) = change_output {
+            pending_transaction.outputs_to_be_received.push(co);
+        }
+
+        self.db
+            .write(WriteOperation::Insert(DbKeyValuePair::PendingTransactionOutputs(
+                tx_id,
+                Box::new(pending_transaction),
+            )))?;
+
+        Ok(())
+    }
+
+    /// when a pending transaction is cancelled the encumbered outputs are moved back to the `unspent_outputs`
+    /// collection.
+    pub fn cancel_pending_transaction_outputs(&mut self, tx_id: TxId) -> Result<(), OutputManagerStorageError> {
+        let pending_tx = fetch!(self, tx_id, PendingTransactionOutputs)?;
+
+        for o in pending_tx.outputs_to_be_spent.iter() {
+            self.db.write(WriteOperation::Insert(DbKeyValuePair::UnspentOutput(
+                o.spending_key.clone(),
+                Box::new(o.clone()),
+            )))?;
+        }
+
+        Ok(())
+    }
+
+    /// This method is check all pending transactions to see if any are older that the provided duration. If they are
+    /// they will be cancelled.
+    pub fn timeout_pending_transaction_outputs(&mut self, period: Duration) -> Result<(), OutputManagerStorageError> {
```

Comment:
Again, I think all this logic should be implemented by the backend. Just say what you want to happen, and the backend will do it.

Reply:
:+1: